### PR TITLE
Sdit 851 improve get /api/offenders/{offender no}/non association details

### PR DESCRIPTION
**Pull request by steverendell**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderNonAssociationDetail.java [hunk 1]
```diff
@@ -113,9 +143,5 @@ public static class Pk implements Serializable {
     public Optional<String> getAgencyDescription() {
         return Optional.ofNullable(offenderBooking.getLocation()).map(AgencyLocation::getDescription);
     }
-
-    public Optional<String> getAssignedLivingUnitDescription() {
-        return Optional.ofNullable(offenderBooking.getAssignedLivingUnit()).map(AgencyInternalLocation::getDescription);
-    }
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderNonAssociationDetail.java:119]: unused

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderNonAssociationDetailRepository.java [hunk 2]
```diff
@@ -1,10 +1,14 @@
...
 import org.springframework.data.repository.CrudRepository;
 import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderNonAssociationDetail;
 
 import java.util.List;
 
 public interface OffenderNonAssociationDetailRepository extends CrudRepository<OffenderNonAssociationDetail, OffenderNonAssociationDetail.Pk> {
-    List<OffenderNonAssociationDetail> findAllByOffenderBooking_BookingId(final long bookingId);
+
+    @EntityGraph(type = EntityGraphType.FETCH, value = "non-association-details")
+    List<OffenderNonAssociationDetail> findAllByOffenderBooking_Offender_NomsIdOrderByEffectiveDateAsc(final String nomisId);
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderNonAssociationDetailRepository.java:13]: this method was previously unused

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsService.java [hunk 3]
```diff
@@ -59,23 +58,23 @@ public OffenderNonAssociationDetails retrieve(final Long bookingId) {
...
-            .flatMap(Collection::stream)
-            .map(this::transform)
-            .toList();
+        final List<OffenderNonAssociationDetail> nonAssociations
+            = offenderNonAssociationDetailRepository.findAllByOffenderBooking_Offender_NomsIdOrderByEffectiveDateAsc(offenderNo);
 
         log.debug("'{}' non-association(s) found for offender '{}'", nonAssociations.size(), offenderNo);
 
+        final var latestBooking = bookingRepository.findWithDetailsByOffenderNomsIdAndBookingSequence(offenderNo, 1)
+            .orElseThrow(EntityNotFoundException.withMessage("Offender no %s not found.", offenderNo));
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderNonAssociationsService.java:67]: I had to split this into 2 repository calls

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderNonAssociationDetail.java [hunk 4]
```diff
@@ -35,6 +37,34 @@
...
 @IdClass(OffenderNonAssociationDetail.Pk.class)
+@NamedEntityGraph(
+    name = "non-association-details",
+    attributeNodes = {
+        @NamedAttributeNode("nonAssociationReason"),
+        @NamedAttributeNode("nonAssociationType"),
+        @NamedAttributeNode("recipNonAssociationReason"),
+        @NamedAttributeNode(value = "nonAssociation", subgraph = "non-association"),
+        @NamedAttributeNode("nsOffender"),
+        // @NamedAttributeNode(value = "bookings", subgraph = "booking-details"),
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderNonAssociationDetail.java:48]: I guess you can delete this line?

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderNonAssociationDetail.java (hunk 4 above)
Comment (steverendell) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderNonAssociationDetail.java:48]: will do next time

**Pull request merged**